runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowRunner.java

```diff
@@ -898,8 +898,19 @@ protected RunnerApi.Pipeline applySdkEnvironmentOverrides(
           throw new RuntimeException("Error parsing environment docker payload.", e);
         }
         String containerImage = dockerPayload.getContainerImage();
+        boolean updated = false;
         for (int i = 0; i < overrides.length; i += 2) {
           containerImage = containerImage.replaceAll(overrides[i], overrides[i + 1]);
+          if (!containerImage.equals(dockerPayload.getContainerImage())) {
+            updated = true;
+          }
+        }
+        if (containerImage.startsWith("apache/beam")
```

Comment:
Does this update all images, including java and go ones? Shouldn't that be done in the service, not in the SDK?

Reply:
It updates foreign images including Python and Go but not a default Java image set by current SDK environment.

I think this can be done by the service too but I think it's safer and has more flexibility when we configure the appropriate URLs from the SDK. Usually we don't encounter this problem since we know the runner by the time we convert pipeline object to pipeline proto. However, for multi-language pipelines, we don't know the runner when we expand external transforms in the expansion service (we just assume open-source runners). We need to update the environment for external transforms when the information about the runner is available.

Reply:
This information is maintained in the SDK side so probably we should do the overriding in the SDK side as well: https://github.com/apache/beam/blob/ead9777d3c1e220604fd12643e40a932631ffe99/runners/google-cloud-dataflow-java/build.gradle#L59

For the primary SDK, we already should be doing this during job submission.

Reply:
BTW Heejong, shouldn't the containers at "apache/beam" path be available and work for Dataflow for released Beam versions even though that's not the preferred option for Dataflow ?

Reply:
I think we don't allow non gcr containers on purpose (minimizing dependency on external service and reducing the point of failures). I couldn't find any published SLA for docker hub service so have no idea how reliable it is. One possible alternative is that we could provide docker hub mirror for official beam images on gcr and configure it as a default container registry instead of external docker hub.

Reply:
Transform expansion should never depend on the runner; if it does, that's a bug. 

We should definitely provide a docker hub mirror for official beam images on gcr and configure it as a default container registry instead of external docker hub; I don't know how hard that is to set up. 

It still feels a bit sketchy for this code to live in each SDK as it seems to be a service/runner issue, but I suppose DataflowRunner is "part" of the runner. 

Also, are we sure that for every `apache/beamX` there is always a corresponding `gcr.io/cloud-dataflow/v1beta3/beamX`, even for dev containers and such? (Specifically, isn't there a gap between when Beam releases happen and the dataflow dev containers are available?)

Reply:
It's in DataflowRunner.java so technically this is a part of the runner. We override the container during job submission for Dataflow (not during expansion).

> We should definitely provide a docker hub mirror for official beam images on gcr and configure it as a default container registry instead of external docker hub

If we want to do this, I would consider this and orthogonal change, not a blocker for x-lang releases. I think it took weeks last time Emily tried to perform such a migration. For now I think we should override the container base to what Dataflow prefers at job submission. We already do this for all non-x-lang Java pipelines and all Python pipelines.

Reply:
This is guaranteed and done during the Beam release process. We also copy all RC containers to the same GCR locations so release validation can be done without updating the container URL as well.